host/app/lib/card-api.gts

```diff
@@ -498,48 +505,77 @@ function getFields<T extends Card>(card: T, onlyComputeds = false): { [P in keyo
+  function getCachedComponent(target: object, property: string, makeComponent: () => ComponentLike<{ Args: never, Blocks: never }>) {
+    let component = componentCache.get(target)?.get(property);
+    if (!component) {
+      component = makeComponent();
+      let targetCache = componentCache.get(target);
+      if (!targetCache) {
+        targetCache = new Map();
+        componentCache.set(target, targetCache);
+      }
+      targetCache.set(property, component);
+    }
+    return component;
+  }
+
   return new Proxy(target, {
     get(target, property, received) {
       if (typeof property === 'symbol' || model == null) {
         // don't handle symbols or nulls
         return Reflect.get(target, property, received);
       }
-      let field = getField(model.constructor, property);
-      if (!field) {
+      let containingModel = modelPath.length === 0 ? model : get(model, modelPath.join('.'));
+      if (!containingModel) {
+        // don't handle nulls and undefineds
+        return Reflect.get(target, property, received);
+      }
+      let _field = getField(containingModel.constructor, property);
+      if (!_field) {
         // field doesn't exist, fall back to normal property access behavior
         return Reflect.get(target, property, received);
       }
-      let innerModel = (model as any)[property];
-      defaultFormat = isFieldComputed(model.constructor, property) ? 'embedded' : defaultFormat;
+      let field = _field;
+      let fieldPath = [...modelPath, property];
+      let innerModel = get(model, fieldPath.join('.'));
+      defaultFormat = isFieldComputed(containingModel.constructor, property) ? 'embedded' : defaultFormat;
 
-      if (isFieldContainsMany(model.constructor, property) && defaultFormat === 'edit') {
+      if (isFieldContainsMany(containingModel.constructor, property) && defaultFormat === 'edit') {
         if (isBaseCard in innerModel) {
           throw new Error('Cannot edit containsMany composite field');
         }
-        let setters = innerModel.map((_el: any, i: number) => makeSetter(model, property, i));
+        let setters = innerModel.map((_el: any, i: number) => makeSetter(model, [ property, String(i) ]));
         let fieldName = property; // to get around lint error
+
+        // TODO the cached contains many editor component is not re-rendering when the cached
+        // containsMany field is mutated. Let's revisit this after the contains many component
+        // uses delegation for rendering its interior fields
         return class ContainsManyEditorTemplate extends GlimmerComponent {
           <template>
             <ContainsManyEditor
-              @model={{model}}
+              @model={{containingModel}}
               @items={{innerModel}}
               @fieldName={{fieldName}}
               @setters={{setters}}
             />
           </template>
         };
```

Comment:
Note that we should revist this. In order to not collide with @burieberry, I'm gonna hold off on this for now